internal/resource.go

```diff
@@ -1,140 +1,263 @@
 // Copyright (c) 2023 Cisco Systems, Inc. and its affiliates
 // All rights reserved.
-
 package internal
 
 import (
 	"context"
 	"crypto/sha256"
 	"encoding/hex"
 	"fmt"
+	"io/ioutil"
+	"net/http"
 	"net/url"
 	"os"
 	"path"
 	"path/filepath"
-
-	"github.com/rs/zerolog/log"
+	"strings"
 
 	"github.com/carlmjohnson/requests"
+	"github.com/rs/zerolog/log"
 )
 
-// Resource represents an external resource to be downloaded.
 type Resource struct {
 	Urls      []string
 	Integrity string
 	Tags      []string `toml:",omitempty"`
 	Filename  string   `toml:",omitempty"`
+	CacheUri  string   `toml:",omitempty"`
 }
 
 func NewResourceFromUrl(urls []string, algo string, tags []string, filename string) (*Resource, error) {
 	if len(urls) < 1 {
 		return nil, fmt.Errorf("empty url list")
 	}
-	url := urls[0]
 	ctx := context.Background()
-	path, err := GetUrltoTempFile(url, ctx)
+
+	path, err := GetUrltoTempFile(urls[0], ctx)
 	if err != nil {
-		return nil, fmt.Errorf("failed to get url: %s", err)
+		return nil, err
 	}
 	defer os.Remove(path)
+
 	integrity, err := getIntegrityFromFile(path, algo)
 	if err != nil {
-		return nil, fmt.Errorf("failed to compute ressource integrity: %s", err)
+		return nil, err
 	}
-	return &Resource{Urls: urls, Integrity: integrity, Tags: tags, Filename: filename}, nil
+
+	return &Resource{
+		Urls:      urls,
+		Integrity: integrity,
+		Tags:      tags,
+		Filename:  filename,
+	}, nil
 }
 
-// getUrl downloads the given resource and returns the path to it.
 func getUrl(u string, fileName string, ctx context.Context) (string, error) {
-	_, err := url.Parse(u)
-	if err != nil {
-		return "", fmt.Errorf("invalid url '%s': %s", u, err)
+	if _, err := url.Parse(u); err != nil {
+		return "", fmt.Errorf("failed to download '%s': invalid URL", u)
 	}
-	log.Debug().Str("URL", u).Msg("Downloading")
-	err = requests.
-		URL(u).
+
+	req := requests.URL(u).
 		Header("Accept", "*/*").
-		ToFile(fileName).
-		Fetch(ctx)
+		ToFile(fileName)
+
+	err := req.Fetch(ctx)
```

Comment:
But... why are you splitting this into two steps if you're not doing anything else with `req`?